chirp-impl/src/main/java/sample/chirper/chirp/impl/JpaReadSideImpl.java

```diff
@@ -0,0 +1,124 @@
+        return new ReadSideHandlerBuilder<Event>() {
+            private PMap<Class<? extends Event>, BiConsumer<EntityManager, ? extends Event>> eventHandlers =
+                    HashTreePMap.empty();
+
+            @Override
+            public <E extends Event> ReadSideHandlerBuilder<Event> setEventHandler(Class<E> eventClass, BiConsumer<EntityManager, E> handler) {
+                eventHandlers = eventHandlers.plus(eventClass, handler);
+                return this;
+            }
+
+            @Override
+            public ReadSideHandler<Event> build() {
+                return new JpaReadSideHandler<>(readSideId, /*globalPrepare, prepare,*/ eventHandlers);
+            }
+        };
+    }
+
+    private class JpaReadSideHandler<Event extends AggregateEvent<Event>> extends ReadSideHandler<Event> {
+        private final String readSideId;
+        private final PMap<Class<? extends Event>, BiConsumer<EntityManager, ? extends Event>> eventHandlers;
+
+        private volatile AggregateEventTag<Event> tag;
+        private volatile JdbcOffsetDao offsetDao;
+
+        private JpaReadSideHandler(
+                String readSideId,
+                PMap<Class<? extends Event>, BiConsumer<EntityManager, ? extends Event>> eventHandlers) {
+            this.readSideId = readSideId;
+            this.eventHandlers = eventHandlers;
+        }
+
+        @Override
+        public CompletionStage<Offset> prepare(AggregateEventTag<Event> tag) {
+            this.tag = tag;
+            return FutureConverters.toJava(offsetStore.prepare(readSideId, tag.tag()))
+                    .thenApply(offsetDao -> {
+                        this.offsetDao = offsetDao;
+                        return offsetDao.loadedOffset();
+                    });
+        }
+
+        @Override
+        public Flow<Pair<Event, Offset>, Done, ?> handle() {
+            return Flow.<Pair<Event, Offset>>create().mapAsync(1, eventAndOffset -> {
+                Event event = eventAndOffset.first();
+                Offset offset = eventAndOffset.second();
+                @SuppressWarnings("unchecked") Class<Event> eventClass = (Class<Event>) event.getClass();
+                @SuppressWarnings("unchecked") BiConsumer<EntityManager, Event> eventHandler =
+                        (BiConsumer<EntityManager, Event>) eventHandlers.get(eventClass);
+                if (eventHandler != null) {
+                    return jpa.withTransaction(entityManager -> {
+                        eventHandler.accept(entityManager, event);
+                        updateOffset(entityManager, offset);
+                        return Done.getInstance();
+                    });
+                } else {
+                    /*
+                        if (log.isDebugEnabled)
+                          log.debug("Unhandled event [{}]", eventClass.getName())
+                     */
+                    return CompletableFuture.completedFuture(Done.getInstance());
```

Comment:
Is the wink about the commented-out logging code? That was copied from the JDBC read side handler... obviously this is all still missing logging.